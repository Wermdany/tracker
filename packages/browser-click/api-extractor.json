{
  "extends": "../../api-extractor.json",
  "mainEntryPointFilePath": "./dist/packages/browser-click/src/index.d.ts",
  "dtsRollup": {
    "publicTrimmedFilePath": "./dist/browser-click.d.ts"
  },
  "compiler": {
    "overrideTsconfig": {
      "compilerOptions": {
        "baseUrl": "packages/browser-click/dist/",
        "paths": {
          "@tracing/*": ["packages/*/src"]
        }
      },
      "include": ["global.d.ts", "packages/*/src/**/*.ts"],
      "exclude": ["test-utils/*"]
    }
  }
}
