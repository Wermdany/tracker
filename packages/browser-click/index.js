"use strict";

if (process.env.NODE_ENV === "production") {
  module.exports = require("./dist/browser-click.cjs.prod.js");
} else {
  module.exports = require("./dist/browser-click.cjs.js");
}
