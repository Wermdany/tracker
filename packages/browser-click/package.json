{
  "name": "@tracing/browser-click",
  "version": "0.0.3",
  "private": false,
  "lib": true,
  "description": "@tracing/browser-click",
  "main": "index.js",
  "module": "dist/browser-click.esm-bundler.js",
  "types": "dist/browser-click.d.ts",
  "sideEffects": false,
  "author": "wermdany <https://github.com/wermdany>",
  "license": "MIT",
  "files": [
    "dist",
    "index.js"
  ],
  "buildOptions": {
    "name": "TracingBrowserClick",
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/wermdany/tracing.git",
    "directory": "packages/browser-click"
  },
  "bugs": {
    "url": "https://github.com/wermdany/tracing/issues"
  },
  "homepage": "https://github.com/wermdany/tracing/tree/master/packages/browser-click#readme",
  "keywords": [
    "tracker",
    "report",
    "collect",
    "FE",
    "web",
    "browser"
  ],
  "dependencies": {
    "@tracing/core": "0.0.3",
    "@tracing/shared": "0.0.3"
  }
}
