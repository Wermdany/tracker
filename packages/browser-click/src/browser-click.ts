import type { TracingPlugin, Logger, TracingCore } from "@tracing/core";

import {
  getElementTagName,
  getAttributeNames,
  isElement,
  getElementClassName,
  getElementSelector,
  getElementPath
} from "@tracing/shared";

export interface BrowserClickPluginConfig {
  document: HTMLElement;
  watchElement: Array<keyof HTMLElementTagNameMap>;
  watchAttrs: string[];
  watchLevel: number;
  genRecord: (target: HTMLElement) => Record<string, any>;
}

export function BrowserClickPlugin(inputConfig?: Partial<BrowserClickPluginConfig>): TracingPlugin {
  const config: BrowserClickPluginConfig = {
    document: document.body,
    watchElement: ["button", "a", "input", "textarea"],
    watchAttrs: ["auto-watch-browser-click"],
    watchLevel: 1,
    genRecord: defaultGenRecord,
    ...inputConfig
  };

  let core: TracingCore;

  const elementIsWatch = (element: HTMLElement, level: number): HTMLElement | undefined => {
    const tagName = getElementTagName(element) as keyof HTMLElementTagNameMap;

    const attrs = getAttributeNames(element);

    if (config.watchElement.includes(tagName) || config.watchAttrs.some(attr => attrs.includes(attr))) {
      return element;
    }
    if (level > 0) {
      return elementIsWatch(element.parentElement!, --level);
    }
  };

  const watch = (logger: Logger) => {
    const dom = config.document;

    if (!dom || !isElement(dom)) {
      __DEV__ && logger.error(`watch document is not Element`);
      return;
    }

    dom.addEventListener("click", watchFun);
  };

  const watchFun = (event: MouseEvent) => {
    const element = event.target as HTMLElement;

    const watchEl = elementIsWatch(element, config.watchLevel);

    if (!watchEl) {
      return;
    }

    const record = config.genRecord(watchEl);

    core.report(BrowserClickEvent, record);
  };

  const unWatch = () => {
    config.document.removeEventListener("click", watchFun);
  };

  return {
    name: "tracing:browser-click",
    init(ctx) {
      core = ctx;
      watch(this.logger);
    },
    destroy() {
      unWatch();
    }
  };
}

export function defaultGenRecord(target: HTMLElement) {
  const tagName = getElementTagName(target);

  const globalRecord = {
    elementTagName: tagName,
    elementClassName: getElementClassName(target),
    elementSelector: getElementSelector(target),
    elementPath: getElementPath(target)
  };

  if (tagName == "a") {
    return { href: (target as HTMLAnchorElement).href, ...globalRecord };
  }
  return globalRecord;
}

export const BrowserClickEvent = "browser-click";
