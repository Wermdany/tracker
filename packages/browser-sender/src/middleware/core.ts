import type { BaseSenderHandle } from "../base";
import type { BatchSenderConfig } from "./batch";
import type { ErrorRetryConfig } from "./retry";

export function compose<R>(...funcs: Function[]): (...args: any[]) => R;
export function compose(...funcs: Function[]) {
  if (funcs.length === 0) {
    return <T>(arg: T) => arg;
  }

  if (funcs.length === 1) {
    return funcs[0];
  }

  return funcs.reduce(
    (a, b) =>
      (...args: any) =>
        a(b(...args))
  );
}

export type Middleware<Base extends BaseSenderHandle = BaseSenderHandle> = (
  request: Base
) => (...args: Parameters<Base>) => void;

export type MiddlewareDestroy = (...args: any[]) => void | Promise<void>;

export type MiddlewareApi<Config extends Record<string, any> = {}> = (config: Partial<Config>) => {
  middleware: Middleware;
  destroy: MiddlewareDestroy;
};

export interface MiddlewareConfig extends Partial<ErrorRetryConfig>, Partial<BatchSenderConfig> {
  middleware: MiddlewareApi[];
}

export function useMiddleware<Ins extends BaseSenderHandle>(instance: Ins, config: MiddlewareConfig) {
  const { destroy, middleware } = config.middleware.reduce(
    (callback, current) => {
      const { middleware, destroy } = current(config);
      callback.destroy.push(destroy);
      callback.middleware.push(middleware);
      return callback;
    },
    { middleware: [], destroy: [] } as {
      middleware: Middleware[];
      destroy: MiddlewareDestroy[];
    }
  );

  return {
    instance: compose<Ins>(...middleware)(instance),
    destroy() {
      return Promise.all(destroy.map(fun => fun()));
    }
  };
}
