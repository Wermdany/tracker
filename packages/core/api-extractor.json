{
  "extends": "../../api-extractor.json",
  "mainEntryPointFilePath": "./dist/packages/<unscopedPackageName>/src/index.d.ts",
  "dtsRollup": {
    "publicTrimmedFilePath": "./dist/<unscopedPackageName>.d.ts"
  },
  "compiler": {
    "overrideTsconfig": {
      "compilerOptions": {
        "baseUrl": "packages/core/dist/",
        "paths": {
          "@collect/*": ["packages/*/src"]
        }
      },
      "include": ["global.d.ts", "packages/*/src/**/*.ts"]
    }
  }
}
