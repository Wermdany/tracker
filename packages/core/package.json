{
  "name": "@tracing/core",
  "version": "0.0.3",
  "private": false,
  "lib": true,
  "description": "@tracing/core",
  "main": "index.js",
  "module": "dist/core.esm-bundler.js",
  "types": "dist/core.d.ts",
  "sideEffects": false,
  "author": "wermdany <https://github.com/wermdany>",
  "license": "ISC",
  "files": [
    "dist",
    "index.js"
  ],
  "buildOptions": {
    "name": "TracingCore",
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/wermdany/tracing.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/wermdany/tracing/issues"
  },
  "homepage": "https://github.com/wermdany/tracing/tree/master/packages/core#readme",
  "keywords": [
    "tracker",
    "report",
    "collect",
    "FE",
    "web",
    "browser"
  ],
  "dependencies": {
    "@tracing/shared": "0.0.3"
  }
}
