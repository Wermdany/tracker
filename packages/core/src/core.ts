import type { TracingCoreConfig } from "./config";
import type { Store } from "./store";

import { createStore } from "./store";
import { PluginDriver } from "./plugin";
import { createInitConfig } from "./config";
import { createLogger } from "./logger";

export class TracingCore {
  [key: string]: any;

  private initialized = false;
  private destroyed = false;

  private pluginDriver;
  private config: TracingCoreConfig;

  public logger;
  public header!: Store<"header">;

  constructor(config: Partial<TracingCoreConfig>) {
    this.config = createInitConfig(config);
    this.logger = createLogger("core", this.config);

    this.pluginDriver = PluginDriver(this.config.plugins.flat(), this.config, this.logger);

    this.pluginDriver.hookSequentialSync("setup", [this.config]);
  }

  public init() {
    if (this.initialized) {
      __DEV__ && this.logger.warn('"init" has been initialized, please do not call again');
      return;
    }

    this.header = createStore("header", {});

    this.initialized = true;
    this.pluginDriver.hookSequentialSync("init", [this]);
  }

  public report(event: string, record: Record<string, any>, meta?: unknown) {
    if (!this.initialized) {
      __DEV__ && this.logger.warn('"TracingCore" Not initialized, please execute after initialization');
      return;
    }

    if (this.destroyed) {
      __DEV__ && this.logger.warn('"TracingCore" has been destroyed and you can\'t do anything now');
      return;
    }

    this.build(event, record, meta);
  }

  private build(event: string, record: Record<string, any>, meta?: unknown) {
    const build = this.pluginDriver.hookChainMergeSync("build", [event, record, meta], {});

    this.send(event, build);
  }

  private send(event: string, build: Record<string, any>) {
    const isBail = this.pluginDriver.hookBail("beforeSend", [event, build]);
    if (isBail === false) return;

    if (__DEV__ && this.config.isLogger && this.config.sendLog) {
      console.log(JSON.stringify(build, null, 2));
    }

    this.pluginDriver.hookBail("send", [event, build]);
  }

  public async destroy() {
    await this.pluginDriver.hookParallel("beforeDestroy", [this]);
    this.pluginDriver.hookSequentialSync("destroy", [this]);

    this.header.clear();

    this.destroyed = true;
  }
}
