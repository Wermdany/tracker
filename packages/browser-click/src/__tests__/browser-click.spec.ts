import { TracingCore } from "@tracing/core";

import { BrowserClickPlugin, defaultGenRecord } from "../browser-click";

import * as modules from "..";

function createTestWrapper() {
  const wrapper = document.createElement("div");
  const div = document.createElement("div");
  const button = document.createElement("button");
  const span = document.createElement("span");
  button.appendChild(span);
  wrapper.appendChild(div);
  wrapper.appendChild(button);

  return { wrapper, div, button, span };
}

describe("browser-click-plugin", () => {
  let all: ReturnType<typeof createTestWrapper>;

  let tc: TracingCore;

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    all = createTestWrapper();

    tc = new TracingCore({
      plugins: [
        BrowserClickPlugin({
          document: all.wrapper
        })
      ],
      isLogger: false
    });

    jest.spyOn(tc, "report");
    jest.spyOn(all.wrapper, "addEventListener");
    jest.spyOn(all.wrapper, "removeEventListener");

    tc.init();
  });

  it("should addEventListener to wrapper", () => {
    expect(all.wrapper.addEventListener).toBeCalledTimes(1);
  });

  it("should watch button", () => {
    all.button.click();
    expect(tc.report).toBeCalled();
  });

  it("should not watch div", () => {
    all.div.click();
    expect(tc.report).not.toBeCalled();
  });

  it("should watch div by level", () => {
    all.span.click();
    expect(tc.report).toBeCalled();
  });

  it("should watch div by attribute", () => {
    all.div.setAttribute("auto-watch-browser-click", "true");
    all.div.click();
    expect(tc.report).toBeCalled();
  });

  it("should removeEventListener to wrapper", async () => {
    await tc.destroy();
    expect(all.wrapper.removeEventListener).toBeCalledTimes(1);
  });

  it("should browser-click output modules", () => {
    expect(modules).toMatchSnapshot();
  });
});

describe("browser-click -> defaultGenRecord", () => {
  it("should use anchor", () => {
    const anchor = document.createElement("a");
    anchor.href = "https://baidu.com/";
    anchor.className = "a";

    expect(defaultGenRecord(anchor)).toEqual({
      elementClassName: "a",
      elementPath: "a.a",
      elementSelector: "a.a",
      elementTagName: "a",
      href: "https://baidu.com/"
    });
  });

  it("should use div", () => {
    const div = document.createElement("div");

    div.className = "a";

    expect(defaultGenRecord(div)).toEqual({
      elementClassName: "a",
      elementPath: "div.a",
      elementSelector: "div.a",
      elementTagName: "div"
    });
  });
});
