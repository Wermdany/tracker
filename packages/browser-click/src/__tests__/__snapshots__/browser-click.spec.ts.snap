// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser-click-plugin should browser-click output modules 1`] = `
{
  "BrowserClickEvent": "browser-click",
  "BrowserClickPlugin": [Function],
  "defaultGenRecord": [Function],
}
`;
